### ISSUE TITLE
wrong path for mksquashfs (squashfs-tools)

### ISSUE THREAD

**Issue by A33a** [apptainer/singularity#1078]
## Version of Singularity:

2.4

### Expected behavior

Build a container.

### Actual behavior

Outputs:
```
frontend:31 16:10:21 ~/sjupyter > singularity build sjupyter.simg docker://a33a/sjupyter
ERROR: You must install squashfs-tools to build images
ABORT: Aborting with RETVAL=255
```

### Steps to reproduce behavior

I use shared filesystem to run the singularity and the squashfs-tools is installed there at /sharedfs/singularity/bin/mksquashfs, right in the bin derectory of singularity executable. $PATH variable contains that path, but when I try to build the image, it doesn't see it. 

What pathes build command uses? Thank you in advance.

P.S. I haven't noticed such behaviour in Singularity version 2.3.2. Doesn't it use squashfs-tools for bootsraping or does it create a "writable" image if squashfs-tools are not found? In version 2.4 build.exec checks for mksquashfs right in the beginning and there is no way to create any other kind of container.

**Comment by DrDaveD** [comment]
I noticed this too, when writing my response in #940.  The problem is that /usr/libexec/singularity/cli/build.exec uses 'which mksquashfs' and there's no which command.  It should use singularity_which instead.

**Comment by A33a** [comment]
@DrDaveD Thanks for an answer. Can you explain please why is there no which? Doesn't it run in the user's shell? There is the which in the system that I am using, does it utilize other binaries? Thank you in advance.

**Comment by DrDaveD** [comment]
I'm not sure what shell you're using, but build.exec uses /bin/bash and it is not a builtin in bash.

**Comment by GodloveD** [comment]
Hi @A33a!  Sorry you ran into this error.  

@DrDaveD is right.  `which` could be causing your problem.  

But there is another issue too.  Singularity explicitly sets the path [here](https://github.com/singularityware/singularity/blob/master/bin/singularity.in#L40).  This is a security measure to prevent a user from slipping an untrusted binary on the path and executing it with elevated privs.  

So `mksquashfs` needs to be installed in a standard system location to be found and used by Singularity.

**Comment by A33a** [comment]
Hi @GodloveD! No problem, I am trying to make the application better :) Now I see that the problem was caused by PATH, because as soon as I moved `mksquashfs` to `/usr/local/bin` it was found (but another error occurred, #1079. It might not be caused by the current issue, though). 

Maybe it is better to somehow select an additional PATH at the configuration step? Or does that option already exist? The point is that Singularity is more focused on shared environment, it would be convenient to use shared fs for the 3rd party software. 

And thank you once again for your efforts :)

**Comment by GodloveD** [comment]
> Maybe it is better to somehow select an additional PATH at the configuration step?

I think that is a great idea.  Actually now that I am thinking about it, this could probably go in `singularity.conf`.  It's safe to put things like this in there because Singularity checks ownership/perms before applying config.

**Comment by dtrudg** [comment]
Closing this as believe fixed via merged #1080 and #1094 - let us know if there is any remaining issue. Thanks.
